File: frontend/src/mocks/data/syncLogs.ts
/**
 * Mock Sync Logs Data
 * TASK-FE-P5-001: MSW Mock Server Setup
 *
 * Historical sync operation logs for data sources.
 */

export interface MockSyncLog {
  id: string;
  data_source: {
    id: string;
    name: string;
  };
  sync_type: 'scheduled' | 'manual' | 'initial';
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | 'cancelled';
  celery_task_id: string | null;
  started_at: string;
  completed_at: string | null;
  duration_seconds: number | null;
  records_processed: number;
  records_created: number;
  records_updated: number;
  records_skipped: number;
  records_failed: number;
  error_message: string | null;
  error_details: Array<{
    record_id: string | null;
    field: string | null;
    message: string;
  }> | null;
  metadata: {
    file_name: string | null;
    file_size_bytes: number | null;
    api_calls_made: number | null;
    triggered_by: string | null;
  } | null;
  created_at: string;
}

// Data sources reference
const dataSources = {
  EPA: {
    id: '550e8400-e29b-41d4-a716-446655440001',
    name: 'EPA GHG Emission Factors Hub',
  },
  DEFRA: {
    id: '550e8400-e29b-41d4-a716-446655440002',
    name: 'DEFRA Conversion Factors',
  },
  EXIOBASE: {
    id: '550e8400-e29b-41d4-a716-446655440003',
    name: 'Exiobase',
  },
};

// Generate UUID
function generateUUID(): string {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    const v = c === 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

// Generate sync logs
function generateSyncLogs(): MockSyncLog[] {
  const logs: MockSyncLog[] = [];
  const now = new Date();

  // EPA syncs - successful history
  for (let i = 0; i < 10; i++) {
    const startDate = new Date(now.getTime() - (i * 14 + 1) * 24 * 60 * 60 * 1000);
    const duration = 300 + Math.floor(Math.random() * 60);
    const endDate = new Date(startDate.getTime() + duration * 1000);

    logs.push({
      id: generateUUID(),
      data_source: dataSources.EPA,
      sync_type: 'scheduled',
      status: 'completed',
      celery_task_id: `celery-task-epa-${generateUUID().slice(0, 8)}`,
      started_at: startDate.toISOString(),
      completed_at: endDate.toISOString(),
      duration_seconds: duration,
      records_processed: 280 + Math.floor(Math.random() * 20),
      records_created: Math.floor(Math.random() * 20),
      records_updated: Math.floor(Math.random() * 15),
      records_skipped: 250 + Math.floor(Math.random() * 20),
      records_failed: 0,
      error_message: null,
      error_details: null,
      metadata: {
        file_name: `ghg_emission_factors_${2023 - Math.floor(i / 2)}.xlsx`,
        file_size_bytes: 1500000 + Math.floor(Math.random() * 100000),
        api_calls_made: null,
        triggered_by: 'celery_beat',
      },
      created_at: startDate.toISOString(),
    });
  }

  // DEFRA syncs - mostly successful with a few minor issues
  for (let i = 0; i < 10; i++) {
    const startDate = new Date(now.getTime() - (i * 14 + 2) * 24 * 60 * 60 * 1000);
    const duration = 480 + Math.floor(Math.random() * 120);
    const endDate = new Date(startDate.getTime() + duration * 1000);
    const recordsFailed = i === 3 ? 5 : i === 7 ? 2 : 0;

    logs.push({
      id: generateUUID(),
      data_source: dataSources.DEFRA,
      sync_type: i === 5 ? 'manual' : 'scheduled',
      status: 'completed',
      celery_task_id: `celery-task-defra-${generateUUID().slice(0, 8)}`,
      started_at: startDate.toISOString(),
      completed_at: endDate.toISOString(),
      duration_seconds: duration,
      records_processed: 370 + Math.floor(Math.random() * 30),
      records_created: 10 + Math.floor(Math.random() * 15),
      records_updated: 15 + Math.floor(Math.random() * 20),
      records_skipped: 340 + Math.floor(Math.random() * 20),
      records_failed: recordsFailed,
      error_message: recordsFailed > 0 ? 'Some records failed validation' : null,
      error_details: recordsFailed > 0
        ? [
            {
              record_id: 'DEFRA-2023-ERR-001',
              field: 'co2e_factor',
              message: 'Value out of expected range',
            },
          ]
        : null,
      metadata: {
        file_name: `defra_conversion_factors_${2024 - Math.floor(i / 3)}.xlsx`,
        file_size_bytes: 2500000 + Math.floor(Math.random() * 200000),
        api_calls_made: null,
        triggered_by: i === 5 ? 'admin_user@example.com' : 'celery_beat',
      },
      created_at: startDate.toISOString(),
    });
  }

  // EXIOBASE syncs - some failures
  for (let i = 0; i < 8; i++) {
    const startDate = new Date(now.getTime() - (i * 30 + 5) * 24 * 60 * 60 * 1000);
    const isFailed = i === 0 || i === 2;
    const duration = isFailed ? 900 : 1800 + Math.floor(Math.random() * 600);
    const endDate = new Date(startDate.getTime() + duration * 1000);

    logs.push({
      id: generateUUID(),
      data_source: dataSources.EXIOBASE,
      sync_type: i === 4 ? 'manual' : 'scheduled',
      status: isFailed ? 'failed' : 'completed',
      celery_task_id: `celery-task-exio-${generateUUID().slice(0, 8)}`,
      started_at: startDate.toISOString(),
      completed_at: endDate.toISOString(),
      duration_seconds: duration,
      records_processed: isFailed ? 500 : 1200 + Math.floor(Math.random() * 100),
      records_created: isFailed ? 0 : 20 + Math.floor(Math.random() * 30),
      records_updated: isFailed ? 0 : 50 + Math.floor(Math.random() * 50),
      records_skipped: isFailed ? 0 : 1100 + Math.floor(Math.random() * 50),
      records_failed: isFailed ? 500 : Math.floor(Math.random() * 5),
      error_message: isFailed ? 'Connection timeout while downloading file' : null,
      error_details: isFailed
        ? [
            {
              record_id: null,
              field: null,
              message: 'HTTP 504: Gateway timeout after 5 retries',
            },
          ]
        : null,
      metadata: {
        file_name: 'exiobase_v3.8.2.zip',
        file_size_bytes: isFailed ? null : 85000000,
        api_calls_made: isFailed ? 5 : 1,
        triggered_by: i === 4 ? 'admin_user@example.com' : 'celery_beat',
      },
      created_at: startDate.toISOString(),
    });
  }

  // Add one in-progress sync
  logs.push({
    id: generateUUID(),
    data_source: dataSources.EPA,
    sync_type: 'manual',
    status: 'in_progress',
    celery_task_id: `celery-task-epa-live-${generateUUID().slice(0, 8)}`,
    started_at: new Date(now.getTime() - 2 * 60 * 1000).toISOString(), // 2 minutes ago
    completed_at: null,
    duration_seconds: null,
    records_processed: 145,
    records_created: 3,
    records_updated: 2,
    records_skipped: 140,
    records_failed: 0,
    error_message: null,
    error_details: null,
    metadata: {
      file_name: 'ghg_emission_factors_2024.xlsx',
      file_size_bytes: 1650000,
      api_calls_made: null,
      triggered_by: 'test_admin@example.com',
    },
    created_at: new Date(now.getTime() - 2 * 60 * 1000).toISOString(),
  });

  // Sort by started_at descending
  logs.sort((a, b) => new Date(b.started_at).getTime() - new Date(a.started_at).getTime());

  return logs;
}

export const mockSyncLogs: MockSyncLog[] = generateSyncLogs();

// Filter helper for sync logs
export function filterSyncLogs(
  logs: MockSyncLog[],
  filters: {
    data_source_id?: string;
    status?: string;
    sync_type?: string;
    start_date?: string;
    end_date?: string;
    has_errors?: boolean;
    sort_by?: string;
    sort_order?: 'asc' | 'desc';
  }
): MockSyncLog[] {
  let filtered = [...logs];

  if (filters.data_source_id) {
    filtered = filtered.filter((l) => l.data_source.id === filters.data_source_id);
  }

  if (filters.status) {
    filtered = filtered.filter((l) => l.status === filters.status);
  }

  if (filters.sync_type) {
    filtered = filtered.filter((l) => l.sync_type === filters.sync_type);
  }

  if (filters.start_date) {
    const startDate = new Date(filters.start_date);
    filtered = filtered.filter((l) => new Date(l.started_at) >= startDate);
  }

  if (filters.end_date) {
    const endDate = new Date(filters.end_date);
    endDate.setHours(23, 59, 59, 999);
    filtered = filtered.filter((l) => new Date(l.started_at) <= endDate);
  }

  if (filters.has_errors) {
    filtered = filtered.filter((l) => l.records_failed > 0);
  }

  // Sorting
  if (filters.sort_by) {
    const order = filters.sort_order === 'asc' ? 1 : -1;
    filtered.sort((a, b) => {
      const aVal = a[filters.sort_by as keyof MockSyncLog];
      const bVal = b[filters.sort_by as keyof MockSyncLog];
      if (aVal === null) return 1;
      if (bVal === null) return -1;
      if (typeof aVal === 'number' && typeof bVal === 'number') {
        return (aVal - bVal) * order;
      }
      if (filters.sort_by === 'started_at' || filters.sort_by === 'completed_at') {
        return (new Date(aVal as string).getTime() - new Date(bVal as string).getTime()) * order;
      }
      return String(aVal).localeCompare(String(bVal)) * order;
    });
  }

  return filtered;
}

// Generate summary statistics
export function generateSyncLogsSummary(logs: MockSyncLog[]) {
  const completedSyncs = logs.filter((l) => l.status === 'completed');
  const failedSyncs = logs.filter((l) => l.status === 'failed');

  const totalRecordsProcessed = logs.reduce((sum, l) => sum + l.records_processed, 0);
  const totalRecordsFailed = logs.reduce((sum, l) => sum + l.records_failed, 0);

  const completedWithDuration = completedSyncs.filter((l) => l.duration_seconds !== null);
  const averageDuration =
    completedWithDuration.length > 0
      ? completedWithDuration.reduce((sum, l) => sum + (l.duration_seconds || 0), 0) /
        completedWithDuration.length
      : 0;

  return {
    total_syncs: logs.length,
    completed_syncs: completedSyncs.length,
    failed_syncs: failedSyncs.length,
    total_records_processed: totalRecordsProcessed,
    total_records_failed: totalRecordsFailed,
    average_duration_seconds: Math.round(averageDuration * 10) / 10,
  };
}

File: backend/api/routes/admin/sync_logs.py
"""
Admin Sync Logs API Routes.

TASK-API-P5-001: Admin Data Sources Endpoints

Endpoints:
- GET /admin/sync-logs - Get sync history with filters
- GET /admin/sync-logs/{id} - Get single sync log details

Contract Reference: phase5-contracts/admin-sync-logs-contract.yaml
"""

import uuid
from datetime import datetime, date, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, and_
from sqlalchemy.orm import Session

from backend.database.connection import get_db
from backend.models import DataSource, DataSyncLog
from backend.schemas.admin import (
    SyncStatusEnum,
    SyncTypeEnum,
    SyncLogSortByEnum,
    SortOrderEnum,
    SyncLogsListResponse,
    SyncLogItem,
    SyncLogDetailResponse,
    SyncLogsSummary,
    SyncLogDataSource,
    ErrorDetail,
)


# ============================================================================
# Router Configuration
# ============================================================================

router = APIRouter(tags=["admin-sync-logs"])


# ============================================================================
# Helper Functions
# ============================================================================


def create_error_response(code: str, message: str, details: Optional[list] = None) -> dict:
    """Create a standardized error response dict."""
    return {
        "error": {
            "code": code,
            "message": message,
            "details": details or [],
        },
        "request_id": f"req_{uuid.uuid4().hex[:12]}",
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def sync_log_to_item(log: DataSyncLog) -> SyncLogItem:
    """Convert a DataSyncLog model to SyncLogItem response."""
    # Calculate duration
    duration_seconds = None
    if log.completed_at and log.started_at:
        duration_seconds = (log.completed_at - log.started_at).total_seconds()

    # Parse error_details if present
    error_details = None
    if log.error_details:
        error_details = [
            ErrorDetail(
                record_id=e.get("record_id"),
                field=e.get("field"),
                message=e.get("message", "Unknown error"),
            )
            for e in log.error_details
        ]

    return SyncLogItem(
        id=log.id,
        data_source=SyncLogDataSource(
            id=log.data_source.id,
            name=log.data_source.name,
        ),
        sync_type=log.sync_type,
        status=log.status,
        celery_task_id=log.celery_task_id,
        started_at=log.started_at,
        completed_at=log.completed_at,
        duration_seconds=duration_seconds,
        records_processed=log.records_processed or 0,
        records_created=log.records_created or 0,
        records_updated=log.records_updated or 0,
        records_skipped=log.records_skipped or 0,
        records_failed=log.records_failed or 0,
        error_message=log.error_message,
        error_details=error_details,
        metadata=log.sync_metadata,
        created_at=log.created_at or log.started_at,
    )


# ============================================================================
# API Endpoints
# ============================================================================


@router.get(
    "/sync-logs",
    response_model=SyncLogsListResponse,
    status_code=status.HTTP_200_OK,
    summary="List sync logs",
    description="Retrieve sync operation history with filtering and pagination.",
)
def list_sync_logs(
    data_source_id: Optional[str] = Query(
        None, description="Filter by specific data source"
    ),
    status_filter: Optional[SyncStatusEnum] = Query(
        None, alias="status", description="Filter by sync status"
    ),
    sync_type: Optional[SyncTypeEnum] = Query(
        None, description="Filter by sync type"
    ),
    start_date: Optional[date] = Query(
        None, description="Filter syncs started on or after this date (YYYY-MM-DD)"
    ),
    end_date: Optional[date] = Query(
        None, description="Filter syncs started on or before this date (YYYY-MM-DD)"
    ),
    has_errors: Optional[bool] = Query(
        None, description="Filter to only syncs with errors (records_failed > 0)"
    ),
    limit: int = Query(
        50, ge=1, le=100, description="Number of results to return"
    ),
    offset: int = Query(
        0, ge=0, description="Number of results to skip"
    ),
    sort_by: SyncLogSortByEnum = Query(
        SyncLogSortByEnum.started_at, description="Sort field"
    ),
    sort_order: SortOrderEnum = Query(
        SortOrderEnum.desc, description="Sort direction"
    ),
    db: Session = Depends(get_db),
) -> SyncLogsListResponse:
    """
    Retrieve sync operation history with filtering and pagination.

    Query Parameters:
    - data_source_id: Filter by specific data source
    - status: Filter by sync status (pending, in_progress, completed, failed, cancelled)
    - sync_type: Filter by sync type (scheduled, manual, initial)
    - start_date: Filter syncs started on or after this date
    - end_date: Filter syncs started on or before this date
    - has_errors: Filter to only syncs with errors (records_failed > 0)
    - limit: Number of results to return (1-100, default: 50)
    - offset: Number of results to skip (default: 0)
    - sort_by: Sort field (started_at, completed_at, records_processed, records_failed)
    - sort_order: Sort direction (asc, desc)

    Returns:
    - items: List of sync log entries
    - total: Total matching sync logs
    - limit/offset: Applied pagination
    - has_more: Whether more results exist
    - summary: Summary of filtered results
    """
    # Validate data_source_id if provided
    if data_source_id:
        data_source = (
            db.query(DataSource)
            .filter(DataSource.id == data_source_id)
            .first()
        )
        if not data_source:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail=create_error_response(
                    code="INVALID_DATA_SOURCE",
                    message="Data source not found",
                    details=[
                        {"field": "data_source_id", "message": f"No data source exists with ID {data_source_id}"}
                    ],
                ),
            )

    # Build base query
    query = db.query(DataSyncLog)

    # Apply filters
    if data_source_id:
        query = query.filter(DataSyncLog.data_source_id == data_source_id)

    if status_filter:
        query = query.filter(DataSyncLog.status == status_filter.value)

    if sync_type:
        query = query.filter(DataSyncLog.sync_type == sync_type.value)

    if start_date:
        start_datetime = datetime.combine(start_date, datetime.min.time())
        query = query.filter(DataSyncLog.started_at >= start_datetime)

    if end_date:
        end_datetime = datetime.combine(end_date, datetime.max.time())
        query = query.filter(DataSyncLog.started_at <= end_datetime)

    if has_errors is True:
        query = query.filter(DataSyncLog.records_failed > 0)
    elif has_errors is False:
        query = query.filter(DataSyncLog.records_failed == 0)

    # Get total count before pagination
    total = query.count()

    # Apply sorting
    sort_column = getattr(DataSyncLog, sort_by.value)
    if sort_order == SortOrderEnum.desc:
        query = query.order_by(sort_column.desc())
    else:
        query = query.order_by(sort_column.asc())

    # Apply pagination
    logs = query.offset(offset).limit(limit).all()

    # Convert to response items
    items = [sync_log_to_item(log) for log in logs]

    # Calculate summary
    # Use the same filtered query for summary stats
    summary_query = db.query(DataSyncLog)
    if data_source_id:
        summary_query = summary_query.filter(DataSyncLog.data_source_id == data_source_id)
    if status_filter:
        summary_query = summary_query.filter(DataSyncLog.status == status_filter.value)
    if sync_type:
        summary_query = summary_query.filter(DataSyncLog.sync_type == sync_type.value)
    if start_date:
        start_datetime = datetime.combine(start_date, datetime.min.time())
        summary_query = summary_query.filter(DataSyncLog.started_at >= start_datetime)
    if end_date:
        end_datetime = datetime.combine(end_date, datetime.max.time())
        summary_query = summary_query.filter(DataSyncLog.started_at <= end_datetime)
    if has_errors is True:
        summary_query = summary_query.filter(DataSyncLog.records_failed > 0)
    elif has_errors is False:
        summary_query = summary_query.filter(DataSyncLog.records_failed == 0)

    total_syncs = summary_query.count()

    completed_syncs = summary_query.filter(DataSyncLog.status == "completed").count()

    failed_syncs = summary_query.filter(DataSyncLog.status == "failed").count()

    total_records_processed = (
        db.query(func.sum(DataSyncLog.records_processed))
        .filter(
            *([DataSyncLog.data_source_id == data_source_id] if data_source_id else []),
            *([DataSyncLog.status == status_filter.value] if status_filter else []),
            *([DataSyncLog.sync_type == sync_type.value] if sync_type else []),
        )
        .scalar() or 0
    )

    total_records_failed = (
        db.query(func.sum(DataSyncLog.records_failed))
        .filter(
            *([DataSyncLog.data_source_id == data_source_id] if data_source_id else []),
            *([DataSyncLog.status == status_filter.value] if status_filter else []),
            *([DataSyncLog.sync_type == sync_type.value] if sync_type else []),
        )
        .scalar() or 0
    )

    # Calculate average duration for completed syncs
    completed_logs = (
        summary_query
        .filter(
            DataSyncLog.status == "completed",
            DataSyncLog.completed_at.isnot(None),
        )
        .all()
    )

    average_duration = None
    if completed_logs:
        durations = [
            (log.completed_at - log.started_at).total_seconds()
            for log in completed_logs
            if log.completed_at and log.started_at
        ]
        if durations:
            average_duration = sum(durations) / len(durations)

    summary = SyncLogsSummary(
        total_syncs=total_syncs,
        completed_syncs=completed_syncs,
        failed_syncs=failed_syncs,
        total_records_processed=total_records_processed,
        total_records_failed=total_records_failed,
        average_duration_seconds=average_duration,
    )

    return SyncLogsListResponse(
        items=items,
        total=total,
        limit=limit,
        offset=offset,
        has_more=total > offset + limit,
        summary=summary,
    )


@router.get(
    "/sync-logs/{sync_log_id}",
    response_model=SyncLogDetailResponse,
    status_code=status.HTTP_200_OK,
    summary="Get sync log details",
    description="Get detailed information for a specific sync log entry.",
    responses={
        404: {"description": "Sync log not found"},
    },
)
def get_sync_log(
    sync_log_id: str,
    db: Session = Depends(get_db),
) -> SyncLogDetailResponse:
    """
    Get detailed information for a specific sync log entry.

    Path Parameters:
    - sync_log_id: Unique sync log identifier

    Returns:
    - Sync log details with full error information

    Raises:
    - 404: Sync log not found
    """
    # Query sync log
    log = (
        db.query(DataSyncLog)
        .filter(DataSyncLog.id == sync_log_id)
        .first()
    )

    if not log:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=create_error_response(
                code="NOT_FOUND",
                message="Sync log not found",
                details=[
                    {"field": "id", "message": f"No sync log exists with ID {sync_log_id}"}
                ],
            ),
        )

    # Calculate duration
    duration_seconds = None
    if log.completed_at and log.started_at:
        duration_seconds = (log.completed_at - log.started_at).total_seconds()

    # Parse error_details if present
    error_details = None
    if log.error_details:
        error_details = [
            ErrorDetail(
                record_id=e.get("record_id"),
                field=e.get("field"),
                message=e.get("message", "Unknown error"),
            )
            for e in log.error_details
        ]

    return SyncLogDetailResponse(
        id=log.id,
        data_source=SyncLogDataSource(
            id=log.data_source.id,
            name=log.data_source.name,
        ),
        sync_type=log.sync_type,
        status=log.status,
        celery_task_id=log.celery_task_id,
        started_at=log.started_at,
        completed_at=log.completed_at,
        duration_seconds=duration_seconds,
        records_processed=log.records_processed or 0,
        records_created=log.records_created or 0,
        records_updated=log.records_updated or 0,
        records_skipped=log.records_skipped or 0,
        records_failed=log.records_failed or 0,
        error_message=log.error_message,
        error_details=error_details,
        metadata=log.sync_metadata,
        created_at=log.created_at or log.started_at,
    )
